client/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,933 @@
+			return nil, result
+		}
+		if unwrapResp == nil || unwrapResp.Auth == nil || unwrapResp.Auth.ClientToken == "" {
+			result = multierror.Append(result, fmt.Errorf("failed to unwrap the token for task %q", taskName))
+			return nil, result
+		}
+
+		// Append the unwrapped token to the return value
+		unwrappedTokens[taskName] = unwrapResp.Auth.ClientToken
+	}
+
+	return unwrappedTokens, nil
+}
+
+// GetConsulACL creates a vault API client and reads from vault a consul ACL
+// token used by the task.
+func (c *vaultClient) GetConsulACL(token, path string) (*vaultapi.Secret, error) {
+	if token == "" {
+		return nil, fmt.Errorf("missing token")
+	}
+	if path == "" {
+		return nil, fmt.Errorf("missing consul ACL token vault path")
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// Use the token supplied to interact with vault
+	c.client.SetToken(token)
+
+	// Restore the token in client to VaultClient's token
+	defer c.client.SetToken(c.config.Token)
+
+	// Read the consul ACL token and return the secret directly
+	return c.client.Logical().Read(path)
+}
+
+// RenewToken renews the supplied token and adds it to the min-heap so that it
+// is renewed periodically by the renewal loop. Any error returned during
+// renewal will be written to a buffered channel and the channel is returned
+// instead of an actual error. This helps the caller be notified of a renewal
+// failure asynchronously for appropriate actions to be taken.
+func (c *vaultClient) RenewToken(token string, increment int) <-chan error {
+	// Create a buffered error channel
+	errCh := make(chan error, 1)
+
+	if token == "" {
+		errCh <- fmt.Errorf("missing token")
+		return errCh
+	}
+	if increment < 1 {
+		errCh <- fmt.Errorf("increment cannot be less than 1")
+		return errCh
+	}
+
+	// Create a renewal request and indicate that the identifier in the
+	// request is a token and not a lease
+	renewalReq := &vaultClientRenewalRequest{
+		errCh:     errCh,
+		id:        token,
+		isToken:   true,
+		increment: increment,
+	}
+
+	// Perform the renewal of the token and send any error to the dedicated
+	// error channel.
+	if err := c.renew(renewalReq); err != nil {
+		errCh <- err
+	}
+
+	return errCh
+}
+
+// RenewLease renews the supplied lease identifier for a supplied duration and
+// adds it to the min-heap so that it gets renewed periodically by the renewal
+// loop. Any error returned during renewal will be written to a buffered
+// channel and the channel is returned instead of an actual error. This helps
+// the caller be notified of a renewal failure asynchronously for appropriate
+// actions to be taken.
+func (c *vaultClient) RenewLease(leaseId string, increment int) <-chan error {
+	c.logger.Printf("[DEBUG] client/vaultclient: renewing lease %q", leaseId)
+	// Create a buffered error channel
+	errCh := make(chan error, 1)
+
+	if leaseId == "" {
+		errCh <- fmt.Errorf("missing lease ID")
+		return errCh
+	}
+
+	if increment < 1 {
```

Comment:
I wonder if we can make a special case number like -1 which means renew immediately so that you can figure out the lease duration. This will be useful when the client restores as it will only have the lease_id. Thoughts?